File: neet_code_all_cpp/sliding_window/num_of_subarray_size_k_avg_threshold.cpp
#include <iostream>
#include <vector>
#include <deque>

class Solution {
    public:
        // TLE(RustだとACした)
        int numOfSubarrays(std::vector<int>& arr, int k, int threshold) {
            int result = 0;

            for(int i = 0; i < arr.size() - k + 1; ++i) {
                int sub_array_sum = 0;
                for(int j = i; j < i + k; ++j) {
                    sub_array_sum += arr[j];
                }

                if(sub_array_sum / k >= threshold) {
                    ++result;
                }
            } 

            return result;
        }

        // AC
        // O(N)
        int numOfSubarrays2(std::vector<int>& arr, int k, int threshold) {
            int result = 0;
            int sub_array_sum = 0;

            int i = 0;
            while(i < k) {
                sub_array_sum += arr[i];
                ++i;
            }
            if(sub_array_sum / k >= threshold) {
                ++result;
            }

            while(i < arr.size()) {
                sub_array_sum -= arr[i - k];
                sub_array_sum += arr[i];

                if(sub_array_sum / k >= threshold) {
                    ++result;
                }
                ++i;
            }

            return result;
        }
};

// 模範解答
class SolutionAns {
    public:
        int numOfSubarrays(std::vector<int>& arr, int k, int threshold) {
            int sum = 0;
            int n = (int)arr.size();
            for(int i = 0; i < k; ++i) {
                sum += arr[i];
            }

            int left = 0;
            int right = k;
            int result = 0;
            if(sum / k >= threshold) ++result;
            while(right < n) {
                sum -= arr[left++];
                sum += arr[right++];
                if(sum /k >= threshold) ++result;
            }

            return result;
        }
};

int main(void) {
    using Case = std::tuple<std::vector<int>, int, int>;

    Case case_1 = {{2, 2, 2, 2, 5, 5, 5, 8}, 3, 4};
    // => 3
    Case case_2 = {{11, 13, 17, 23, 29, 31, 7, 5, 2, 3}, 3, 6};
    // => 6

    Solution s_1;

    std::cout << s_1.numOfSubarrays(std::get<0>(case_1), std::get<1>(case_1), std::get<2>(case_1)) << std::endl;
    std::cout << s_1.numOfSubarrays(std::get<0>(case_2), std::get<1>(case_2), std::get<2>(case_2)) << std::endl;

    std::cout << s_1.numOfSubarrays2(std::get<0>(case_1), std::get<1>(case_1), std::get<2>(case_1)) << std::endl;
    std::cout << s_1.numOfSubarrays2(std::get<0>(case_2), std::get<1>(case_2), std::get<2>(case_2)) << std::endl;

    SolutionAns s_ans;

    std::cout << s_ans.numOfSubarrays(std::get<0>(case_1), std::get<1>(case_1), std::get<2>(case_1)) << std::endl;
    std::cout << s_ans.numOfSubarrays(std::get<0>(case_2), std::get<1>(case_2), std::get<2>(case_2)) << std::endl;
}

File: neet_code_all/src/sliding_window/num_of_subarray_size_k_avg_threshold.rs
struct Solution {}
impl Solution {
    // O(kN) 実質O(N^2)
    // AC
    pub fn num_of_subarray(arr: Vec<i32>, k: i32, threshold: i32) -> i32 {
        let mut result = 0;

        for i in 0..(arr.len() as i32 - k + 1) {
            let mut sub_array_sum = 0;
            for j in i..(i + k) {
                sub_array_sum += arr[j as usize];
            }

            if sub_array_sum / k >= threshold {
                result += 1;
            }
        }

        result
    }

    //AC
    //O(N)
    pub fn num_of_subarray_2(arr: Vec<i32>, k: i32, threshold: i32) -> i32 {
        let mut result = 0;
        let mut sub_array_sum = 0;

        let mut i = 0;
        while i < k {
            sub_array_sum += arr[i as usize];
            i += 1;
        }
        if sub_array_sum / k >= threshold {
            result += 1;
        }

        while i < arr.len() as i32 {
            sub_array_sum -= arr[(i - k) as usize];
            sub_array_sum += arr[i as usize];

            if sub_array_sum / k >= threshold {
                result += 1;
            }

            i += 1;
        }

        result
    }
}

// 長さがkで平均値がthreshold以下の部分列の個数

fn main() {
    let case_1 = (vec![2, 2, 2, 2, 5, 5, 5, 8], 3, 4);
    // => 3
    let case_2 = (vec![11, 13, 17, 23, 29, 31, 7, 5, 2, 3], 3, 6);
    // => 6

    println!(
        "case_1: {}",
        Solution::num_of_subarray(case_1.0.clone(), case_1.1, case_1.2)
    );
    println!(
        "case_2: {}",
        Solution::num_of_subarray(case_2.0.clone(), case_2.1, case_2.2)
    );

    println!(
        "case_1: {}",
        Solution::num_of_subarray_2(case_1.0.clone(), case_1.1, case_1.2)
    );
    println!(
        "case_2: {}",
        Solution::num_of_subarray_2(case_2.0.clone(), case_2.1, case_2.2)
    );
}
